NPE when hash join

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```sql
CREATE TABLE `t2` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
CREATE TABLE `t` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
alter table t2 set tiflash replica 1;
alter table t set tiflash replica 1;
SELECT count(*) FROM test.t2 left outer join test.t on if(test.t2.a,null,null);
```
```
[2023/12/22 10:43:27.319 +08:00] [ERROR] [BaseDaemon.cpp:367] [########################################] [source=BaseDaemon] [thread_id=940]
[2023/12/22 10:43:27.319 +08:00] [ERROR] [BaseDaemon.cpp:368] ["(from thread 19) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=940]
[2023/12/22 10:43:27.319 +08:00] [ERROR] [BaseDaemon.cpp:396] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=940]
[2023/12/22 10:43:27.319 +08:00] [ERROR] [BaseDaemon.cpp:404] ["Access: read."] [source=BaseDaemon] [thread_id=940]
[2023/12/22 10:43:27.319 +08:00] [ERROR] [BaseDaemon.cpp:413] ["Address not mapped to object."] [source=BaseDaemon] [thread_id=940]
[2023/12/22 10:43:27.336 +08:00] [ERROR] [BaseDaemon.cpp:560] ["\n       0x77b4801\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+125519873]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:211\n  0x7f9950c67db0\t<unknown symbol> [libc.so.6+347568]\n       0x7efd703\tDB::recordFilteredRows(DB::Block const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, COWPtr<DB::IColumn>::immutable_ptr<DB::IColumn>&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul> const*&) [tiflash+133158659]\n                \tdbms/src/Interpreters/JoinUtils.cpp:88\n       0x7f0ae0e\tDB::ProbeProcessInfo::prepareForCrossProbe(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::ASTTableJoin::Kind, DB::ASTTableJoin::Strictness, DB::Block const&, unsigned long, DB::CrossProbeMode, unsigned long) [tiflash+133213710]\n                \tdbms/src/Interpreters/ProbeProcessInfo.cpp:122\n       0x7c6f98b\tDB::Join::joinBlock(DB::ProbeProcessInfo&, bool) const [tiflash+130480523]\n                \tdbms/src/Interpreters/Join.cpp:2143\n       0x88ee3a4\tDB::HashJoinProbeTransformOp::transformHeaderImpl(DB::Block&) [tiflash+143582116]\n                \tdbms/src/Operators/HashJoinProbeTransformOp.cpp:61\n       0x89bc75f\tDB::PipelineExecBuilder::appendTransformOp(std::__1::unique_ptr<DB::TransformOp, std::__1::default_delete<DB::TransformOp> >&&) [tiflash+144426847]\n                \tdbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp:28\n       0x8a64805\tDB::PhysicalJoinProbe::buildPipelineExecGroupImpl(DB::PipelineExecutorContext&, DB::PipelineExecGroupBuilder&, DB::Context&, unsigned long) [tiflash+145115141]\n                \tdbms/src/Flash/Planner/Plans/PhysicalJoinProbe.cpp:50\n       0x89f45ee\tDB::PhysicalPlanNode::buildPipelineExecGroup(DB::PipelineExecutorContext&, DB::PipelineExecGroupBuilder&, DB::Context&, unsigned long) [tiflash+144655854]\n                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:105\n       0x89b1aaf\tDB::Pipeline::buildExecGroup(DB::PipelineExecutorContext&, DB::Context&, unsigned long) [tiflash+144382639]\n                \tdbms/src/Flash/Pipeline/Pipeline.cpp:202\n       0x89cc532\tDB::PlainPipelineEvent::scheduleImpl() [tiflash+144491826]\n                \tdbms/src/Flash/Pipeline/Schedule/Events/PlainPipelineEvent.cpp:24\n       0x89c56f2\tDB::Event::schedule() [tiflash+144463602]\n                \tdbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:132\n       0x89c678c\tDB::Event::finish() [tiflash+144467852]\n                \tdbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:193\n       0x89d52b0\tDB::EventTask::finalizeImpl() [tiflash+144528048]\n                \tdbms/src/Flash/Pipeline/Schedule/Tasks/EventTask.cpp:41\n       0x89db3e2\tDB::Task::finalize() [tiflash+144552930]\n                \tdbms/src/Flash/Pipeline/Schedule/Tasks/Task.cpp:162\n       0x1eb407b\tDB::TaskThreadPool<DB::CPUImpl>::loop(unsigned long) [tiflash+32194683]\n                \tdbms/src/Flash/Pipeline/Schedule/ThreadPool/TaskThreadPool.cpp:61\n       0x1eb47c6\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::TaskThreadPool<DB::CPUImpl>::*)(unsigned long), DB::TaskThreadPool<DB::CPUImpl>*, unsigned long> >(void*) [tiflash+32196550]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7f9950cb2802\tstart_thread [libc.so.6+653314]"] [source=BaseDaemon] [thread_id=940]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
master

--- a/dbms/src/Interpreters/JoinUtils.cpp
+++ b/dbms/src/Interpreters/JoinUtils.cpp
@@ -49,8 +49,6 @@ void recordFilteredRows(
     if (filter_column.empty())
         return;
     auto column = block.getByName(filter_column).column;
-    if (column->isColumnConst())
-        column = column->convertToFullColumnIfConst();
     if unlikely (column->onlyNull())
     {
         if (!null_map_holder)
@@ -68,6 +66,9 @@ void recordFilteredRows(
         null_map = &static_cast<const ColumnUInt8 &>(*null_map_holder).getData();
         return;
     }
+    
+    if (column->isColumnConst())
+        column = column->convertToFullColumnIfConst();
 
     const PaddedPODArray<UInt8> * column_data;
     if (column->isColumnNullable())

--- a/dbms/src/Interpreters/JoinUtils.cpp
+++ b/dbms/src/Interpreters/JoinUtils.cpp
@@ -63,6 +63,7 @@ void recordFilteredRows(
             PaddedPODArray<UInt8> & mutable_null_map = static_cast<ColumnUInt8 &>(*mutable_null_map_holder).getData();
             mutable_null_map.resize_fill(0);
             mutable_null_map.resize_fill(column->size(), 1);
+            null_map_holder = std::move(mutable_null_map_holder);
         }
         null_map = &static_cast<const ColumnUInt8 &>(*null_map_holder).getData();
         return;

--- a/dbms/src/Interpreters/JoinUtils.cpp
+++ b/dbms/src/Interpreters/JoinUtils.cpp
@@ -66,7 +66,7 @@ void recordFilteredRows(
         null_map = &static_cast<const ColumnUInt8 &>(*null_map_holder).getData();
         return;
     }
-    
+
     if (column->isColumnConst())
         column = column->convertToFullColumnIfConst();
 

--- a/dbms/src/Interpreters/JoinUtils.cpp
+++ b/dbms/src/Interpreters/JoinUtils.cpp
@@ -51,6 +51,23 @@ void recordFilteredRows(
     auto column = block.getByName(filter_column).column;
     if (column->isColumnConst())
         column = column->convertToFullColumnIfConst();
+    if unlikely (column->onlyNull())
+    {
+        if (!null_map_holder)
+        {
+            null_map_holder = ColumnVector<UInt8>::create(column->size(), 1);
+        }
+        else
+        {
+            MutableColumnPtr mutable_null_map_holder = (*std::move(null_map_holder)).mutate();
+            PaddedPODArray<UInt8> & mutable_null_map = static_cast<ColumnUInt8 &>(*mutable_null_map_holder).getData();
+            mutable_null_map.resize_fill(0);
+            mutable_null_map.resize_fill(column->size(), 1);
+        }
+        null_map = &static_cast<const ColumnUInt8 &>(*null_map_holder).getData();
+        return;
+    }
+
     const PaddedPODArray<UInt8> * column_data;
     if (column->isColumnNullable())
     {

--- a/tests/fullstack-test/issues/issue_8562.test
+++ b/tests/fullstack-test/issues/issue_8562.test
@@ -17,7 +17,6 @@ mysql> drop table if exists test.t1;
 mysql> CREATE TABLE test.t1 (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
 mysql> drop table if exists test.t2;
 mysql> CREATE TABLE test.t2 (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
-mysql> drop table if exists test.t2;
 mysql> insert into test.t1 values(30, 50);
 mysql> insert into test.t1 values(300, 500);
 mysql> insert into test.t2 values(30, 50);

--- a/tests/fullstack-test/issues/issue_8562.test
+++ b/tests/fullstack-test/issues/issue_8562.test
@@ -29,7 +29,7 @@ mysql> alter table test.t2 set tiflash replica 1;
 func> wait_table test t1
 func> wait_table test t2
 
-mysql> use test; set @@tidb_isolation_read_engines='tiflash'; set @@tidb_enforce_mpp=1; SELECT * FROM test.t2 left outer join test.t on if(test.t2.a,null,null);
+mysql> use test; set @@tidb_isolation_read_engines='tiflash'; set @@tidb_enforce_mpp=1; SELECT * FROM test.t2 left outer join test.t1 on if(test.t2.a,null,null);
 +-----+------+------+------+
 | a   | b    | a    | b    |
 +-----+------+------+------+

--- a/tests/fullstack-test/issues/issue_8562.test
+++ b/tests/fullstack-test/issues/issue_8562.test
@@ -14,9 +14,9 @@
 
 # Preparation.
 mysql> drop table if exists test.t1;
-mysql> CREATE TABLE `test.t1` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> CREATE TABLE test.t1 (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
 mysql> drop table if exists test.t2;
-mysql> CREATE TABLE `test.t2` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> CREATE TABLE test.t2 (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
 mysql> drop table if exists test.t2;
 mysql> insert into test.t1 values(30, 50);
 mysql> insert into test.t1 values(300, 500);

--- a/tests/fullstack-test/issues/issue_8562.test
+++ b/tests/fullstack-test/issues/issue_8562.test
@@ -1,0 +1,42 @@
+# Copyright 2023 PingCAP, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Preparation.
+mysql> drop table if exists test.t1;
+mysql> CREATE TABLE `t1` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> drop table if exists test.t2;
+mysql> CREATE TABLE `t2` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> drop table if exists test.t2;
+mysql> insert into t1 values(30, 50);
+mysql> insert into t1 values(300, 500);
+mysql> insert into t2 values(30, 50);
+mysql> insert into t2 values(300, 500);
+
+mysql> alter table test.t1 set tiflash replica 1;
+mysql> alter table test.t2 set tiflash replica 1;
+
+func> wait_table test t1
+func> wait_table test t2
+
+mysql> use test; set @@tidb_isolation_read_engines='tiflash'; set @@tidb_enforce_mpp=1; SELECT * FROM test.t2 left outer join test.t on if(test.t2.a,null,null);
++-----+------+------+------+
+| a   | b    | a    | b    |
++-----+------+------+------+
+|  30 |   50 | NULL | NULL |
+| 300 |  500 | NULL | NULL |
++-----+------+------+------+
+
+# Clean up.
+mysql> drop table if exists test.t1;
+mysql> drop table if exists test.t2;

--- a/tests/fullstack-test/issues/issue_8562.test
+++ b/tests/fullstack-test/issues/issue_8562.test
@@ -14,14 +14,14 @@
 
 # Preparation.
 mysql> drop table if exists test.t1;
-mysql> CREATE TABLE `t1` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> CREATE TABLE `test.t1` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
 mysql> drop table if exists test.t2;
-mysql> CREATE TABLE `t2` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+mysql> CREATE TABLE `test.t2` (   `a` int(11) NOT NULL,   `b` int(11) DEFAULT NULL,   PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
 mysql> drop table if exists test.t2;
-mysql> insert into t1 values(30, 50);
-mysql> insert into t1 values(300, 500);
-mysql> insert into t2 values(30, 50);
-mysql> insert into t2 values(300, 500);
+mysql> insert into test.t1 values(30, 50);
+mysql> insert into test.t1 values(300, 500);
+mysql> insert into test.t2 values(30, 50);
+mysql> insert into test.t2 values(300, 500);
 
 mysql> alter table test.t1 set tiflash replica 1;
 mysql> alter table test.t2 set tiflash replica 1;

